httpclient/useragent.go

```diff
@@ -38,3 +40,134 @@ func (rt *userAgentRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
 	}
 	return rt.inner.RoundTrip(req)
 }
+
+func UserAgent(version string) *userAgent {
+	return newUserAgent([]*UserAgentProduct{
+		{"HashiCorp", "1.0", ""},
+		{"Terraform", version, "+https://www.terraform.io"},
+	})
+}
+
+type UserAgentProduct struct {
+	Name    string
+	Version string
+	Comment string
+}
+
+func (uap *UserAgentProduct) String() string {
+	var b strings.Builder
+	b.WriteString(uap.Name)
+	if uap.Version != "" {
+		b.WriteString(fmt.Sprintf("/%s", uap.Version))
+	}
+	if uap.Comment != "" {
+		if uap.Name != "" {
```

Comment:
This does not appear to be RFC7231 compliant.

From section 5.5.3:
>  User-Agent = product *( RWS ( product / comment ) )
>
>  The User-Agent field-value consists of one or more product
   identifiers, each followed by zero or more comments

Reply:
@paultyng suggested it is?

I have to admit I'm struggling to interpret the RFC, whilst also correlating all this with [reality](http://www.user-agents.org/).

The [RFC](https://tools.ietf.org/html/rfc7231#section-5.5.3) mentions the following [ABNF](https://tools.ietf.org/html/rfc5234#section-3.2) patterns:

```
User-Agent      = product *( RWS ( product / comment ) )
product         = token ["/" product-version]
product-version = token
```

You are right that it does mention **one or more** product identifiers.

Reply:
I'm inclined to apply the pragmatic approach here, as long as it doesn't hurt RFC compliant input (i.e. as long as valid input is still parsed correctly), which I don't think it does?

Reply:
I agree that the RFC is not entirely clear: there seems to be an inconsistency between the grammar and the text describing it. 

In the ABNF patterns you reproduced, the `User-Agent` rule contains the rule name `product` only once, with no variable repetition. Assuming that the rule name `product` corresponds with "product identifier" in the text, the grammar does not seem to allow for more than one product identifier, making the text "one or more product identifiers" misleading.

However, the example given later in the same section of the RFC seems to show the intent more clearly:
```
User-Agent: CERN-LineMode/2.15 libwww/2.17b3
```

Based on this, I believe the grammar should have been:
```
User-Agent      = 1*(product *( RWS ( product / comment ) ))
product         = token ["/" product-version]
product-version = token
```

Note that neither option for resolving the ambiguity permits `product` to be absent entirely from the `User-Agent` field.

I think the three examples on user-agents.org of the form `( token )` are not spec compliant.

Reply:
I agree with your interpretation, but also I don't mind the pragmatism over strict RFC compliance.

I'm happy to discuss the regexp in the other function, but technically this condition just prevents leading space in (sic non-compliant) string?

I don't have a strong opinion either way. 🤷‍♂️